File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Listeners/onJoin.java
package jp.aoichaan0513.A_TosoGame_Live.Listeners;

import jp.aoichaan0513.A_TosoGame_Live.API.Enums.Advancement;
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.*;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerConfig;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldConfig;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Maps.MapUtility;
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Scoreboard;
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams;
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import jp.aoichaan0513.A_TosoGame_Live.Runnable.RespawnRunnable;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.scoreboard.DisplaySlot;

public class onJoin implements Listener {

    @EventHandler
    public void onJoin(PlayerJoinEvent e) {
        Player p = e.getPlayer();

        e.setJoinMessage(!MainAPI.isHidePlayer(p) ? ChatColor.YELLOW + "-> " + ChatColor.GOLD + p.getName() : "");

        WorldConfig worldConfig = Main.getWorldConfig();

        PlayerConfig playerConfig = PlayerManager.loadConfig(p);
        if (!playerConfig.getAdvancementConfig().hasAdvancement(Advancement.FIRST_JOIN)) {
            playerConfig.getAdvancementConfig().addAdvancement(Advancement.FIRST_JOIN);
            Advancement.FIRST_JOIN.sendMessage(p);
        }

        if (!TosoGameAPI.difficultyMap.containsKey(p.getUniqueId()))
            TosoGameAPI.difficultyMap.put(p.getUniqueId(), WorldManager.Difficulty.NORMAL);

        WorldConfig.DifficultyConfig difficultyConfig = worldConfig.getDifficultyConfig(WorldManager.Difficulty.NORMAL);

        if (TosoGameAPI.isAdmin(p) && MapUtility.getMap() == null)
            p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.WARNING) + "マップの設定がまだ完了していません。\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "マップの設定をした後に\"/map generate\"を実行してください。");

        BossBarManager.showBar(p);
        if (!MoneyManager.hasRate(p))
            MoneyManager.setRate(p, difficultyConfig.getRate());

        Scoreboard.addHidePlayer(p);
        TosoGameAPI.setPotionEffect(p, true);

        MissionManager.reloadBook(p);
        if (MissionManager.isBossBar())
            MissionManager.getBossBar().addPlayer(p);

        if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_PLAYER, p)) {
            Main.opGamePlayerSet.add(p.getUniqueId());

            if (OPGameManager.getOPGame())
                TosoGameAPI.teleport(p, worldConfig.getOPGameLocationConfig().getGOPLocations().values());
        } else if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_JAIL, p)) {
            RespawnRunnable.setCoolTime(p);
            /*
            p.addPotionEffect(new PotionEffect(PotionEffectType.FIRE_RESISTANCE, TosoGameAPI.respawnCoolTimeMap.getOrDefault(p.getUniqueId(), 1) * (difficultyConfig.getRespawnCoolTime() * 20), 1, false, false));
            if (TosoGameAPI.respawnCoolTimeMap.containsKey(p.getUniqueId()))
                TosoGameAPI.respawnCoolTimeMap.replace(p.getUniqueId(), TosoGameAPI.respawnCoolTimeMap.get(p.getUniqueId()), TosoGameAPI.respawnCoolTimeMap.get(p.getUniqueId()) + 1);
            else
                TosoGameAPI.respawnCoolTimeMap.put(p.getUniqueId(), 2);
            */

            TosoGameAPI.teleport(p, worldConfig.getJailLocationConfig().getLocations().values());
        } else if (!Teams.hasJoinedTeams(p)) {
            if (GameManager.isGame()) {
                if (TosoGameAPI.isRes) {
                    Teams.joinTeam(Teams.OnlineTeam.TOSO_PLAYER, p);

                    TosoGameAPI.setItem(WorldManager.GameType.START, p);
                    TosoGameAPI.setPotionEffect(p);

                    Main.opGamePlayerSet.add(p.getUniqueId());

                    if (OPGameManager.getOPGame()) {
                        TosoGameAPI.teleport(p, worldConfig.getOPGameLocationConfig().getGOPLocations().values());
                    } else {
                        TosoGameAPI.teleport(p, worldConfig.getRespawnLocationConfig().getLocations().values());
                    }
                } else {
                    p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "途中参加が禁止のため牢獄に追加されました。");

                    Teams.joinTeam(Teams.OnlineTeam.TOSO_JAIL, p);

                    /*
                    p.addPotionEffect(new PotionEffect(PotionEffectType.FIRE_RESISTANCE, TosoGameAPI.respawnCoolTimeMap.getOrDefault(p.getUniqueId(), 1) * (difficultyConfig.getRespawnCoolTime() * 20), 1, false, false));
                    if (TosoGameAPI.respawnCoolTimeMap.containsKey(p.getUniqueId()))
                        TosoGameAPI.respawnCoolTimeMap.replace(p.getUniqueId(), TosoGameAPI.respawnCoolTimeMap.get(p.getUniqueId()), TosoGameAPI.respawnCoolTimeMap.get(p.getUniqueId()) + 1);
                    else
                        TosoGameAPI.respawnCoolTimeMap.put(p.getUniqueId(), 2);
                    */

                    TosoGameAPI.teleport(p, worldConfig.getJailLocationConfig().getLocations().values());
                }
            } else {
                Teams.joinTeam(Teams.OnlineTeam.TOSO_PLAYER, p);

                TosoGameAPI.setItem(WorldManager.GameType.START, p);
                TosoGameAPI.setPotionEffect(p);

                Main.opGamePlayerSet.add(p.getUniqueId());

                TosoGameAPI.teleport(p, worldConfig.getRespawnLocationConfig().getLocations().values());
            }
        }

        for (Player player : Bukkit.getOnlinePlayers()) {
            TosoGameAPI.showPlayers(player);

            org.bukkit.scoreboard.Scoreboard board = Scoreboard.getBoard(player);

            if (!TosoGameAPI.isAdmin(player)) {
                if (player.getInventory().getItemInMainHand().getType() == Material.BOOK) {
                    ItemMeta itemMeta = player.getInventory().getItemInMainHand().getItemMeta();

                    if (ChatColor.stripColor(itemMeta.getDisplayName()).equals(Main.PHONE_ITEM_NAME)) {
                        board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);
                    } else {
                        board.clearSlot(DisplaySlot.SIDEBAR);
                    }
                } else if (player.getInventory().getItemInOffHand().getType() == Material.BOOK) {
                    ItemMeta itemMeta = player.getInventory().getItemInOffHand().getItemMeta();

                    if (ChatColor.stripColor(itemMeta.getDisplayName()).equals(Main.PHONE_ITEM_NAME)) {
                        board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);
                    } else {
                        board.clearSlot(DisplaySlot.SIDEBAR);
                    }
                } else {
                    board.clearSlot(DisplaySlot.SIDEBAR);
                }
            } else {
                board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);
            }
        }
        p.setPlayerListHeaderFooter("" + ChatColor.RED + ChatColor.BOLD + "Run" + ChatColor.RESET + ChatColor.GRAY + " for " + ChatColor.DARK_RED + ChatColor.BOLD + "Money", "");
        p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "現在のチームは" + ChatColor.BOLD + ChatColor.UNDERLINE + Teams.getJoinedTeam(p).getDisplayName() + ChatColor.RESET + ChatColor.GRAY + "に設定されています。\n" +
                MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "現在の難易度は" + ChatColor.BOLD + ChatColor.UNDERLINE + TosoGameAPI.difficultyMap.get(p.getUniqueId()).getDisplayName() + ChatColor.RESET + ChatColor.GRAY + "に設定されています。");

        org.bukkit.scoreboard.Scoreboard scoreboard = Scoreboard.getBoard(p);
        Teams.setTeamOption(scoreboard);
        p.setScoreboard(scoreboard);

        Teams.setTeamOptions();
        return;
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Listeners/Minecraft/onJoin.kt
package jp.aoichaan0513.A_TosoGame_Live.Listeners.Minecraft

import jp.aoichaan0513.A_TosoGame_Live.API.Advancement.Advancement
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.PrefixType
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.BossBarManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.GameManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.MoneyManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.DifficultyManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerConfig
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldManager.GameType
import jp.aoichaan0513.A_TosoGame_Live.API.Map.MapUtility
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Scoreboard
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams.OnlineTeam
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI
import jp.aoichaan0513.A_TosoGame_Live.Main
import jp.aoichaan0513.A_TosoGame_Live.Mission.HunterZone
import jp.aoichaan0513.A_TosoGame_Live.Mission.MissionManager
import jp.aoichaan0513.A_TosoGame_Live.OPGame.OPGameManager
import jp.aoichaan0513.A_TosoGame_Live.Runnable.RespawnRunnable
import jp.aoichaan0513.A_TosoGame_Live.Utils.isJailTeam
import jp.aoichaan0513.A_TosoGame_Live.Utils.isPlayerGroup
import jp.aoichaan0513.A_TosoGame_Live.Utils.isPlayerTeam
import org.bukkit.ChatColor
import org.bukkit.GameMode
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

class onJoin : Listener {

    @EventHandler
    fun onJoin(e: PlayerJoinEvent) {
        val p = e.player

        e.joinMessage = if (!PlayerManager.loadConfig(p).visibility) "${ChatColor.YELLOW}-> ${ChatColor.GOLD}${p.name}" else ""
        p.sendMessage("${MainAPI.getPrefix(PrefixType.SECONDARY)}リソースパックを適用するには\"${ChatColor.GOLD}${ChatColor.UNDERLINE}/resource${ChatColor.RESET}${ChatColor.GRAY}\"と入力してください。")

        MainAPI.sendDebugMessage(p, """
            ${MainAPI.getPrefix(PrefixType.ERROR)}${ChatColor.DARK_RED}${ChatColor.BOLD}${ChatColor.UNDERLINE}デバッグモードが有効${ChatColor.RESET}${ChatColor.RED}になっています。
            ${MainAPI.getPrefix(PrefixType.SECONDARY)}本番環境では必ずデバッグモードを無効にしてください。
        """.trimIndent())

        val worldConfig = Main.worldConfig
        val playerConfig = PlayerManager.loadConfig(p)

        playerConfig.bookForegroundColor = PlayerConfig.BookForegroundColor.BLACK

        if (playerConfig.advancementConfig.hasAdvancement(Advancement.FIRST_JOIN) && !playerConfig.advancementConfig.hasAdvancement(Advancement.UPDATE_2_0_0))
            Advancement.UPDATE_2_0_0.addAdvancement(p)

        Advancement.FIRST_JOIN.addAdvancement(p)

        val difficulty = playerConfig.difficulty
        if (!DifficultyManager.isDifficulty(p))
            DifficultyManager.setDifficulty(p, difficulty)

        val difficultyConfig = worldConfig.getDifficultyConfig(difficulty)

        if (TosoGameAPI.isAdmin(p) && MapUtility.map == null)
            p.sendMessage("""
                ${MainAPI.getPrefix(PrefixType.WARNING)}マップの設定がまだ完了していません。
                ${MainAPI.getPrefix(PrefixType.SECONDARY)}マップの設定をした後に"${ChatColor.GOLD}${ChatColor.UNDERLINE}/map generate${ChatColor.RESET}${ChatColor.GRAY}"を実行してください。
            """.trimIndent())

        BossBarManager.showBar(p)

        if (!MoneyManager.hasRate(p))
            MoneyManager.setRate(p, difficultyConfig.rate)

        Scoreboard.addHidePlayer(p)
        TosoGameAPI.setPotionEffect(p)

        MissionManager.setBook(p)
        if (MissionManager.isBossBar)
            MissionManager.bossBar!!.addPlayer(p)

        if (HunterZone.internalMissionState == MissionManager.InternalMissionState.END
                && p.isPlayerGroup && !HunterZone.containsCodeSet(p)) {
            p.sendMessage("""
                ${MainAPI.getPrefix(PrefixType.WARNING)}あなたはコードを入力していないため発光しました。
                ${MainAPI.getPrefix(PrefixType.SECONDARY)}解除するには"${ChatColor.UNDERLINE}/code 入手したコード${ChatColor.RESET}${ChatColor.GRAY}"と入力して装置を解除してください。
            """.trimIndent())
            p.addPotionEffect(PotionEffect(PotionEffectType.GLOWING, 200000, 1, false, false))
        }

        if (p.isPlayerTeam) {
            Teams.joinTeam(OnlineTeam.TOSO_PLAYER, p)
            p.gameMode = GameMode.ADVENTURE

            TosoGameAPI.removeOp(p)

            if (OPGameManager.opGameState != OPGameManager.OPGameState.NONE)
                TosoGameAPI.teleport(p, worldConfig.opGameLocationConfig.gOPLocations.values)
            if (HunterZone.containsLeavedSet(p)) {
                TosoGameAPI.teleport(p, worldConfig.respawnLocationConfig.locations.values)
                HunterZone.removeLeavedSet(p)
            }
        } else if (p.isJailTeam) {
            Teams.joinTeam(OnlineTeam.TOSO_JAIL, p)
            p.gameMode = GameMode.ADVENTURE

            TosoGameAPI.setItem(GameType.START, p)
            TosoGameAPI.setPotionEffect(p)
            TosoGameAPI.removeOp(p)

            RespawnRunnable.setAutoTime(p)
            RespawnRunnable.setCoolTime(p)

            TosoGameAPI.teleport(p, worldConfig.jailLocationConfig.locations.values)
        } else if (!Teams.hasJoinedTeams(p)) {
            if (GameManager.isGame()) {
                if (TosoGameAPI.isRespawn) {
                    Teams.joinTeam(OnlineTeam.TOSO_PLAYER, p)
                    p.gameMode = GameMode.ADVENTURE

                    TosoGameAPI.setItem(GameType.START, p)
                    TosoGameAPI.setPotionEffect(p)
                    TosoGameAPI.removeOp(p)

                    if (OPGameManager.opGameState != OPGameManager.OPGameState.NONE)
                        TosoGameAPI.teleport(p, worldConfig.opGameLocationConfig.gOPLocations.values)
                    else
                        TosoGameAPI.teleport(p, worldConfig.respawnLocationConfig.locations.values)
                } else {
                    Teams.joinTeam(OnlineTeam.TOSO_JAIL, p)
                    p.gameMode = GameMode.ADVENTURE

                    TosoGameAPI.setItem(GameType.START, p)
                    TosoGameAPI.setPotionEffect(p)
                    TosoGameAPI.removeOp(p)

                    TosoGameAPI.teleport(p, worldConfig.jailLocationConfig.locations.values)

                    p.sendMessage("${MainAPI.getPrefix(PrefixType.SECONDARY)}途中参加が禁止のため牢獄に追加されました。")
                }
            } else {
                Teams.joinTeam(OnlineTeam.TOSO_PLAYER, p)
                p.gameMode = GameMode.ADVENTURE

                TosoGameAPI.setItem(GameType.START, p)
                TosoGameAPI.setPotionEffect(p)
                TosoGameAPI.removeOp(p)

                TosoGameAPI.teleport(p, worldConfig.respawnLocationConfig.locations.values)
            }
        }

        /*
        for (player in Bukkit.getOnlinePlayers())
            player.setSidebar()
        */

        TosoGameAPI.sendInformationText(p)

        val scoreboard = Scoreboard.getBoard(p)
        Teams.setTeamOption(scoreboard)
        p.scoreboard = scoreboard
        //  Teams.setTeamOptions()
        return
    }
}